topi/python/topi/x86/conv2d.py

```diff
@@ -462,8 +699,212 @@ def traverse(op):
+        s[data_pad].vectorize(dpcii)
+    if cfg['data_pad_compute_location'].val == 2:
+        parallel_axis = s[input_tile].fuse(n, cii)
+        s[data_pad].compute_at(s[input_tile], oh_m)
+        (_, _, _, _, dpcii) = s[data_pad].op.axis
+        s[data_pad].vectorize(dpcii)
+    if cfg['data_pad_compute_location'].val == 3:
+        parallel_axis = s[input_tile].fuse(n, cii, oh_m)
+        s[data_pad].compute_at(s[input_tile], ow_m)
+        (_, _, _, _, dpcii) = s[data_pad].op.axis
+        s[data_pad].vectorize(dpcii)
+
+    # s[input_tile].parallel(parallel_axis)
+    ############################################################
+
+    ############################################################
+    # data_pad
+    # s[data_pad].compute_inline()
+    ############################################################
+
+    ############################################################
+    # transform image
+    n, cii, oh_m, ow_m, eps, nu, ciii = s[V].op.axis
+    r_eps, r_nu = s[V].op.reduce_axis
+
+    s[V].vectorize(ciii)
+    cfg["reorder_V"].apply(s, V, [n, cii, oh_m, ow_m, eps, nu, ciii, r_eps, r_nu])
+
+    cfg.define_annotate("reduce_V", [r_eps, r_nu, eps, nu],
+                        policy='unroll')
+    cfg['reduce_V'].apply(s, V, [r_eps, r_nu, eps, nu], cfg=cfg)
+
+
+    cfg.define_knob('input_tile_compute_location', [0, 1, 2, 3])
+    if cfg['input_tile_compute_location'].val == 0:
+        parallel_axis = s[V].fuse(n)
+    if cfg['input_tile_compute_location'].val == 1:
+        parallel_axis = s[V].fuse(n)
+        s[input_tile].compute_at(s[V], cii)
+    if cfg['input_tile_compute_location'].val == 2:
+        parallel_axis = s[V].fuse(n, cii)
+        s[input_tile].compute_at(s[V], oh_m)
+    if cfg['input_tile_compute_location'].val == 3:
+        parallel_axis = s[V].fuse(n, cii, oh_m)
+        s[input_tile].compute_at(s[V], ow_m)
+
+    # s[V].parallel(parallel_axis)
+    ############################################################
+
+    ############################################################
+    # batch gemm
+    n, coo, oh_m, ow_m, eps, nu, vc = s[M].op.axis
+    cii, ciii = s[M].op.reduce_axis
+    s[M].vectorize(vc)
+
+    cfg["reorder_M"].apply(s, M, [n, coo, oh_m, ow_m, eps, nu, vc, cii, ciii])
+
+    cfg.define_annotate("reduce_M", [eps, nu],
+                        policy='unroll')
+    cfg['reduce_M'].apply(s, M, [eps, nu], cfg=cfg)
+
+    cfg.define_knob('V_compute_location', [0, 1, 2, 3])
+    if cfg['V_compute_location'].val == 0:
+        parallel_axis = s[M].fuse(n)
+    if cfg['V_compute_location'].val == 1:
+        parallel_axis = s[M].fuse(n)
+        s[V].compute_at(s[M], coo)
+    if cfg['V_compute_location'].val == 2:
+        parallel_axis = s[M].fuse(n, coo)
+        s[V].compute_at(s[M], oh_m)
+    if cfg['V_compute_location'].val == 3:
+        parallel_axis = s[M].fuse(n, coo)
+        s[V].compute_at(s[M], ow_m)
+
+    # s[M].parallel(parallel_axis)
```

Comment:
@merrymercy ,

 I believe here we just can ```return``` in case of single thread. 

* Could hint how to check this info right here ? From where to extract thread_num or similar info ?